### ISSUE TITLE
reset_states() failure in a stateful network with initial_states set and training in batch - TypeError: 'NoneType' object is not subscriptable

### ISSUE THREAD

**Issue by manojrege** [keras-team/keras#11148]
Following is the stacktrace - 
```
Traceback (most recent call last):
  File "/usr/local/Cellar/python/3.6.5_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/Cellar/python/3.6.5_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/manoj/Desktop/repos/yane/yane/LSTM/manytomanyLSTM.py", line 137, in <module>
    incremental_train(space)
  File "/Users/manoj/Desktop/repos/yane/yane/LSTM/manytomanyLSTM.py", line 128, in incremental_train
    model.reset_states()
  File "/usr/local/lib/python3.6/site-packages/keras/engine/topology.py", line 1968, in reset_states
    layer.reset_states()
  File "/usr/local/lib/python3.6/site-packages/keras/layers/recurrent.py", line 681, in reset_states
    batch_size = self.input_spec[0].shape[0]        
TypeError: 'NoneType' object is not subscriptable
```

**Comment by manojrege** [comment]
Steps to reproduce - Encoder decoder model from Keras

```encoder_inputs = Input(batch_shape=(32, 1, 1),name='ENCODER_INPUT')
encoder = LSTM(256, return_sequences = True,stateful = True, name='LSTM', return_state = True)
encoder_outputs, state_h, state_c = encoder(encoder_inputs)

encoder_states = [state_h, state_c]

# Set up the decoder, using `encoder_states` as initial state.
decoder_inputs = Input(shape=(1,), batch_shape=(32, tsteps, 1),name='DECODER_INPUT')

decoder = LSTM(256, return_sequences = True, stateful = True, name='LSTM2')
decoder_out = decoder(decoder_inputs, initial_state=encoder_states)
decoder_2 = LSTM(256,return_sequences=False, stateful=True, batch_size=32, name='LSTM3')(decoder_out)
 decoder_outputs = Dense(1,name='DENSE')(decoder_2)

model = Model([encoder_inputs, decoder_inputs], decoder_outputs)
model.compile(loss='mse', optimizer='adam', metrics=['mean_absolute_percentage_error', relative_prediction_error])
for i in range(epochs):
            print('Epoch', i, '/', epochs)
            model.fit([interleaved_trainX, interleaved_trainY], trainY,
                      batch_size=32,
                      verbose=1,
                      epochs=1,
                      shuffle=False)
            model.reset_states()
loss, mape, pe = model.evaluate([interleaved_trainX, interleaved_trainY], trainY, batch_size=32, verbose=1)```

**Comment by manojrege** [comment]
Once the initial_states is removed, the code works. Investigated the issue.  This is the problem. https://github.com/keras-team/keras/blob/40a9ecc684d2c5e7fdfc09f46d0912940b74ab63/keras/layers/recurrent.py#L571

The input_spec is reset to origianl_spec and not full_input_spec

**Comment by huan** [comment]
I ran into this issue today with TF 2.0 preview.

**Comment by bertsky** [comment]
Could this fix be cherrypicked into master?

**Comment by huan** [comment]
@bertsky According to https://github.com/tensorflow/tensorflow/issues/25985#issuecomment-471559617 , this issue should have already fixed by the latest version of TF 2.0

**Comment by bertsky** [comment]
@huan The fix does work. I was merely striving to speed up integrating this in Keras master – before the tf-2 branch gets updated and merged.
